From da4c937f5e9fed5ce3c789b5efc9f229c8f6f717 Mon Sep 17 00:00:00 2001
From: Nicolas LOUBOUTIN <nicolas.louboutin@st.com>
Date: Thu, 6 Aug 2020 09:19:06 +0200
Subject: [PATCH 31/31] ANDROID: scripts: add clang and default reload options
 to merge configs script

1- add -C <triple> option to use clang toolchain
2- add -d option for force default config reload

Signed-off-by: Nicolas LOUBOUTIN <nicolas.louboutin@st.com>
---
 scripts/kconfig/merge_config.sh | 35 +++++++++++++++++++++++++++++++--
 1 file changed, 33 insertions(+), 2 deletions(-)

diff --git a/scripts/kconfig/merge_config.sh b/scripts/kconfig/merge_config.sh
index 63c8565206a4..89e9930bb78a 100755
--- a/scripts/kconfig/merge_config.sh
+++ b/scripts/kconfig/merge_config.sh
@@ -27,6 +27,8 @@ usage() {
 	echo "  -n    use allnoconfig instead of alldefconfig"
 	echo "  -r    list redundant entries when merging fragments"
 	echo "  -y    make builtin have precedence over modules"
+	echo "  -C    <triple> clang build selected"
+	echo "  -d    <config> force default config (several configs possible separated by space)"
 	echo "  -O    dir to put generated output files.  Consider setting \$KCONFIG_CONFIG instead."
 	echo
 	echo "Used prefix: '$CONFIG_PREFIX'. You can redefine it with \$CONFIG_ environment variable."
@@ -36,6 +38,7 @@ RUNMAKE=true
 ALLTARGET=alldefconfig
 WARNREDUN=false
 BUILTIN=false
+CLANG=false
 OUTPUT=.
 CONFIG_PREFIX=${CONFIG_-CONFIG_}
 
@@ -65,6 +68,24 @@ while true; do
 		shift
 		continue
 		;;
+	"-C")
+		CLANG=true
+		CLANG_TRIPLE=$2
+		shift 2
+		;;
+	"-d")
+		FORCE_DEFAULT=true
+		shift
+		CONFIG_DELETE_LIST=""
+		while true; do
+			if [[ "$1" == CONFIG_* ]]; then
+				CONFIG_DELETE_LIST+="$1 "
+				shift
+			else
+				break
+			fi
+		done
+		;;
 	"-O")
 		if [ -d $2 ];then
 			OUTPUT=$(echo $2 | sed 's/\/*$//')
@@ -168,12 +189,22 @@ if [ "$OUTPUT" != "." ] ; then
 	OUTPUT_ARG="O=$OUTPUT"
 fi
 
+# To force default, remove the config from the temporary config file
+if [ $FORCE_DEFAULT = "true" ]; then
+	for config in $CONFIG_DELETE_LIST; do
+		sed -i "/^${config}/d" $TMP_FILE
+		sed -i "/^\# ${config} is not set/d" $TMP_FILE
+	done
+fi
 
 # Use the merged file as the starting point for:
 # alldefconfig: Fills in any missing symbols with Kconfig default
 # allnoconfig: Fills in any missing symbols with # CONFIG_* is not set
-make KCONFIG_ALLCONFIG=$TMP_FILE $OUTPUT_ARG $ALLTARGET
-
+if [ "$CLANG" = "false" ]; then
+	make KCONFIG_ALLCONFIG=$TMP_FILE $OUTPUT_ARG $ALLTARGET
+else
+	make KCONFIG_ALLCONFIG=$TMP_FILE $OUTPUT_ARG CLANG_TRIPLE=$CLANG_TRIPLE CC=clang $ALLTARGET
+fi
 
 # Check all specified config values took (might have missed-dependency issues)
 for CFG in $(sed -n -e "$SED_CONFIG_EXP1" -e "$SED_CONFIG_EXP2" $TMP_FILE); do
-- 
2.17.1

